akka-http-core/src/main/scala/akka/http/impl/engine/server/HttpServerBluePrint.scala

```diff
@@ -89,54 +89,99 @@ private[http] object HttpServerBluePrint {
+      }
 
-      val idle = new InHandler {
-        def onPush(): Unit = grab(in) match {
-          case RequestStart(method, uri, protocol, hdrs, entityCreator, _, _) ⇒
-            val effectiveMethod = if (method == HttpMethods.HEAD && settings.transparentHeadRequests) HttpMethods.GET else method
-            val effectiveHeaders =
-              if (settings.remoteAddressHeader && remoteAddress.isDefined)
-                headers.`Remote-Address`(RemoteAddress(remoteAddress.get)) +: hdrs
-              else hdrs
-
-            val entity = createEntity(entityCreator) withSizeLimit settings.parserSettings.maxContentLength
-            push(out, HttpRequest(effectiveMethod, uri, effectiveHeaders, entity, protocol))
-          case other ⇒
-            throw new IllegalStateException(s"unexpected element of type ${other.getClass}")
+      setIdleHandlers()
+
+      def setIdleHandlers() {
+        setHandler(in, this)
+        setHandler(out, this)
+        if (upstreamPullWaiting) {
+          upstreamPullWaiting = false
+          pull(in)
         }
       }
-      setHandler(in, idle)
+
 
       def createEntity(creator: EntityCreator[RequestOutput, RequestEntity]): RequestEntity =
         creator match {
           case StrictEntityCreator(entity) ⇒ entity
-          case StreamedEntityCreator(creator) ⇒
-            val entitySource = new SubSourceOutlet[RequestOutput]("EntitySource")
-            entitySource.setHandler(new OutHandler {
-              def onPull(): Unit = pull(in)
-            })
-            setHandler(in, new InHandler {
-              def onPush(): Unit = grab(in) match {
-                case MessageEnd ⇒
-                  entitySource.complete()
-                  setHandler(in, idle)
-                case x ⇒ entitySource.push(x)
-              }
-              override def onUpstreamFinish(): Unit = completeStage()
-            })
-            creator(Source.fromGraph(entitySource.source))
+          case StreamedEntityCreator(creator) ⇒ streamRequestEntity(creator)
         }
 
-      setHandler(out, new OutHandler {
-        override def onPull(): Unit = pull(in)
-      })
+      def streamRequestEntity(creator: (Source[ParserOutput.RequestOutput, NotUsed]) => RequestEntity): RequestEntity = {
+        // stream incoming chunks into the request entity until we reach the end of it
+        // and then toggle back to "idle"
+
+        val entitySource = new SubSourceOutlet[RequestOutput]("EntitySource")
+        // optimization: re-use the idle outHandler
+        entitySource.setHandler(this)
+
+        setHandler(in, new InHandler {
+          def onPush(): Unit = {
+            grab(in) match {
+              case MessageEnd ⇒
+                entitySource.complete()
+                setIdleHandlers()
+
+              case x ⇒ entitySource.push(x)
+            }
+          }
+          override def onUpstreamFinish(): Unit = {
+            entitySource.complete()
+            completeStage()
```

Comment:
Ok, added a test that documents this assumption. (I think that http://tools.ietf.org/html/rfc7230#section-4.1 says that you must end chunked stream correctly with a "last-chunk" but could not figure out if the parser handles this already, and maybe the responsibility belongs there anyway)